source/client_channel_handler.c

```diff
@@ -94,11 +89,46 @@ static void s_on_time_to_ping(struct aws_channel_task *channel_task, void *arg,
         s_schedule_ping(connection);
     }
 }
-static int s_packet_handler_connack(
+
+static int s_validate_received_packet_type(
     struct aws_mqtt_client_connection_311_impl *connection,
-    struct aws_byte_cursor message_cursor) {
+    enum aws_mqtt_packet_type packet_type) {
+    { /* BEGIN CRITICAL SECTION */
+        mqtt_connection_lock_synced_data(connection);
+        /* [MQTT-3.2.0-1] The first packet sent from the Server to the Client MUST be a CONNACK Packet */
+        if (connection->synced_data.state == AWS_MQTT_CLIENT_STATE_CONNECTING &&
+            packet_type != AWS_MQTT_PACKET_CONNACK) {
+            mqtt_connection_unlock_synced_data(connection);
+            AWS_LOGF_ERROR(
+                AWS_LS_MQTT_CLIENT,
+                "id=%p: First message received from the server was not a CONNACK. Terminating connection.",
+                (void *)connection);
+            aws_channel_shutdown(connection->slot->channel, AWS_ERROR_MQTT_PROTOCOL_ERROR);
```

Comment:
this error-path calls channel_shutdown() but other error-paths don't

probably simpler to just raise an error here, and let channel_shutdown() get called after aws_mqtt311_decoder_on_bytes_received() fails. Vs calling channel_shutdown() is some places you raise an error, and not others, and then call channel_shutdown() should get called in the end anyway

Reply:
Gonna leave this in since it forces a protocol error and its the original stuff which I don't want to mess with any more than necessary.  But I'm going to do a second shutdown in the call to the decoder from the channel read message process function.